src/details/ArborX_DetailsCartesianGrid.hpp

```diff
@@ -0,0 +1,114 @@
+/****************************************************************************
+ * Copyright (c) 2017-2022 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_DETAILS_CARTESIAN_GRID_HPP
+#define ARBORX_DETAILS_CARTESIAN_GRID_HPP
+
+#include <ArborX_Exception.hpp>
+#include <ArborX_GeometryTraits.hpp>
+
+#include <Kokkos_Macros.hpp>
+
+namespace ArborX::Details
+{
+
+template <class Box>
+struct CartesianGrid
+{
+  static constexpr int DIM = GeometryTraits::dimension<Box>::value;
+
+  Box _bounds;
+  float _h[DIM];
+  size_t _n[DIM];
+
+  CartesianGrid(Box const &bounds, float h)
+      : _bounds(bounds)
+  {
+    for (int d = 0; d < DIM; ++d)
+      _h[d] = h;
+    buildGrid();
+  }
+  CartesianGrid(Box const &bounds, float const h[DIM])
+      : _bounds(bounds)
+  {
+    for (int d = 0; d < DIM; ++d)
+      _h[d] = h[d];
+    buildGrid();
+  }
+
+  template <typename Point, typename Enable = std::enable_if_t<
+                                GeometryTraits::is_point<Point>{}>>
+  KOKKOS_FUNCTION size_t cellIndex(Point const &point) const
+  {
+    static_assert(GeometryTraits::dimension<Point>::value == DIM);
+
+    auto const &min_corner = _bounds.minCorner();
+    size_t s = 0;
+    for (int d = DIM - 1; d >= 0; --d)
+    {
+      int i = std::floor((point[d] - min_corner[d]) / _h[d]);
```

Comment:
```suggestion
      int i = Kokkos::Experimental::floor((point[d] - min_corner[d]) / _h[d]);
```

Reply:
ping

Reply:
Are you suggesting adding `KokkosExt::floor`?